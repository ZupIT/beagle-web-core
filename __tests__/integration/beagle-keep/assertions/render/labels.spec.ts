/*
 * Copyright 2020, 2022 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import setup from '../../backend/routes'
import {
  RenderingResult,
  renderLabelsView,
  getRepeater,
  getViewWithAnEmptyRepeater,
} from '../utils'

/**
 * The view labels is way too similar to the view home, and, since we already tested home
 * thoroughly, we'll only test the final renders and not lifecycle by lifecycle.
 */
describe('Beagle Keep: render labels', () => {
  let renderedTrees: RenderingResult
  setup()
  
  beforeAll(async () => {
    renderedTrees = await renderLabelsView()
  })

  /**
   * Two renders are expected:
   * - first: the view itself (labels), with an empty repeater (no children).
   * - second: the repeater is initialized and calls for a second render of the view labels, now
   * with as many children as elements in its data source.
   */
  it('should do two full renders with no errors or warnings', () => {
    expect(renderedTrees.render.length).toBe(2)
    expect(globalMocks.log).not.toHaveBeenCalled()
  })

  /**
   * Tests the first render of the view labels. The important point here is that nothing under the
   * repeater's template should be processed. The repeater will only have children in the second
   * render of the view.
   */
  it('should render labels for the first time', () => {
    expect(renderedTrees.render[0]).toMatchSnapshot()
  })

  /**
   * Tests the second render of the view labels. This second render is triggered by the component
   * "repeater" via the viewContentManager. Here, the repeater should have content, its children
   * should have been calculated according to the template and data source.
   * 
   * The rest of the tree should be exactly the same as the first render.
   */
  it('should render labels for the second time, now with the list of labels', () => {
    const labels = renderedTrees.render[1]
    const repeater = getRepeater(labels)
    expect(repeater).toMatchSnapshot()
  })

  /**
   * The rest of the tree should be exactly the same as the first render.
   */
  it('second render should be the same as the first, except for the repeater', () => {
    const current = renderedTrees.render[1]
    const previous = renderedTrees.render[0]
    const labelsWithEmptyRepeater = getViewWithAnEmptyRepeater(current)
    expect(JSON.stringify(labelsWithEmptyRepeater)).toEqual(JSON.stringify(previous))
  })
})
