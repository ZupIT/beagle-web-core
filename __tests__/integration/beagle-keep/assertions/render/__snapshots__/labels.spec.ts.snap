// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Beagle Keep: render labels should render labels for the first time 1`] = `
Object {
  "_beagleComponent_": "beagle:container",
  "children": Array [
    Object {
      "_beagleComponent_": "custom:repeater",
      "dataSource": Array [
        Object {
          "color": "#00ff00",
          "id": 0,
          "name": "Personal",
        },
        Object {
          "color": "#0000ff",
          "id": 1,
          "name": "Work",
        },
        Object {
          "color": "#ff0000",
          "id": 2,
          "name": "Urgent",
        },
        Object {
          "color": "#ffff00",
          "id": 3,
          "name": "Supermarket",
        },
      ],
      "id": "label-list",
      "key": "id",
      "style": Object {
        "backgroundColor": "#333",
        "borderRadius": "20px",
        "flexDirection": "column",
        "maxWidth": "500px",
        "padding": "15px",
      },
      "template": Object {
        "_beagleComponent_": "custom:editableLabel",
        "context": Object {
          "id": "isLoading",
          "value": false,
        },
        "isLoading": "@{isLoading}",
        "label": "@{item}",
        "onRemove": Array [
          Object {
            "_beagleAction_": "beagle:condition",
            "condition": "@{isEmpty(item.id)}",
            "onFalse": Array [
              Object {
                "_beagleAction_": "beagle:setContext",
                "contextId": "isLoading",
                "value": true,
              },
              Object {
                "_beagleAction_": "beagle:sendRequest",
                "method": "DELETE",
                "onError": Array [
                  Object {
                    "_beagleAction_": "custom:feedback",
                    "message": "Error while deleting label.",
                    "type": "error",
                  },
                  Object {
                    "_beagleAction_": "beagle:setContext",
                    "contextId": "isLoading",
                    "value": false,
                  },
                ],
                "onSuccess": Array [
                  Object {
                    "_beagleAction_": "beagle:setContext",
                    "contextId": "labels",
                    "path": "current",
                    "value": "@{remove(labels.current, item)}",
                  },
                ],
                "url": "https://keep.beagle.io/label/@{item.id}",
              },
            ],
            "onTrue": Array [
              Object {
                "_beagleAction_": "beagle:setContext",
                "contextId": "labels",
                "path": "new",
              },
            ],
          },
        ],
        "onSave": Array [
          Object {
            "_beagleAction_": "beagle:setContext",
            "contextId": "isLoading",
            "value": true,
          },
          Object {
            "_beagleAction_": "beagle:sendRequest",
            "data": "@{onSave.value}",
            "method": "@{condition(isEmpty(item.id), 'POST', 'PUT')}",
            "onError": Array [
              Object {
                "_beagleAction_": "custom:feedback",
                "message": "@{error.data.message}",
                "type": "error",
              },
              Object {
                "_beagleAction_": "beagle:setContext",
                "contextId": "isLoading",
                "value": false,
              },
            ],
            "onSuccess": Array [
              Object {
                "_beagleAction_": "beagle:condition",
                "condition": "isEmpty(item.id)",
                "onFalse": Array [
                  Object {
                    "_beagleAction_": "beagle:setContext",
                    "contextId": "labels",
                    "path": "current[@{index}]",
                    "value": "@{onSuccess.data}",
                  },
                  Object {
                    "_beagleAction_": "beagle:setContext",
                    "contextId": "isLoading",
                    "value": false,
                  },
                ],
                "onTrue": Array [
                  Object {
                    "_beagleAction_": "beagle:setContext",
                    "contextId": "labels",
                    "value": Object {
                      "current": "@{insert(labels.current, onSuccess.data)}",
                    },
                  },
                ],
              },
            ],
            "url": "https://keep.beagle.io/label",
          },
        ],
      },
    },
    Object {
      "_beagleComponent_": "beagle:container",
      "children": Array [
        Object {
          "_beagleComponent_": "beagle:button",
          "id": "_beagle_1",
          "isSubmit": false,
          "onPress": [Function],
          "text": "Go back",
        },
        Object {
          "_beagleComponent_": "beagle:button",
          "disabled": false,
          "id": "_beagle_2",
          "isSubmit": false,
          "onPress": [Function],
          "text": "Create new label",
        },
      ],
      "id": "container:1",
      "style": Object {
        "color": "#FFF",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
    },
  ],
  "context": Object {
    "id": "labels",
    "value": Object {
      "current": Array [
        Object {
          "color": "#00ff00",
          "id": 0,
          "name": "Personal",
        },
        Object {
          "color": "#0000ff",
          "id": 1,
          "name": "Work",
        },
        Object {
          "color": "#ff0000",
          "id": 2,
          "name": "Urgent",
        },
        Object {
          "color": "#ffff00",
          "id": 3,
          "name": "Supermarket",
        },
      ],
    },
  },
  "id": "container:0",
  "style": Object {
    "alignItems": "center",
    "backgroundColor": "#111",
    "color": "#FFF",
    "flex": 1,
    "height": "100%",
    "justifyContent": "center",
  },
  "version": "Beagle Keep v1.0",
}
`;

exports[`Beagle Keep: render labels should render labels for the second time, now with the list of labels 1`] = `
Object {
  "_beagleComponent_": "custom:repeater",
  "children": Array [
    Object {
      "_beagleComponent_": "custom:editableLabel",
      "_implicitContexts_": Array [
        Object {
          "id": "item",
          "value": Object {
            "color": "#00ff00",
            "id": 0,
            "name": "Personal",
          },
        },
        Object {
          "id": "index",
          "value": 0,
        },
      ],
      "context": Object {
        "id": "isLoading",
        "value": false,
      },
      "id": "repeater:label-list:0:0",
      "isLoading": false,
      "key": 0,
      "label": Object {
        "color": "#00ff00",
        "id": 0,
        "name": "Personal",
      },
      "onRemove": [Function],
      "onSave": [Function],
    },
    Object {
      "_beagleComponent_": "custom:editableLabel",
      "_implicitContexts_": Array [
        Object {
          "id": "item",
          "value": Object {
            "color": "#0000ff",
            "id": 1,
            "name": "Work",
          },
        },
        Object {
          "id": "index",
          "value": 1,
        },
      ],
      "context": Object {
        "id": "isLoading",
        "value": false,
      },
      "id": "repeater:label-list:1:0",
      "isLoading": false,
      "key": 1,
      "label": Object {
        "color": "#0000ff",
        "id": 1,
        "name": "Work",
      },
      "onRemove": [Function],
      "onSave": [Function],
    },
    Object {
      "_beagleComponent_": "custom:editableLabel",
      "_implicitContexts_": Array [
        Object {
          "id": "item",
          "value": Object {
            "color": "#ff0000",
            "id": 2,
            "name": "Urgent",
          },
        },
        Object {
          "id": "index",
          "value": 2,
        },
      ],
      "context": Object {
        "id": "isLoading",
        "value": false,
      },
      "id": "repeater:label-list:2:0",
      "isLoading": false,
      "key": 2,
      "label": Object {
        "color": "#ff0000",
        "id": 2,
        "name": "Urgent",
      },
      "onRemove": [Function],
      "onSave": [Function],
    },
    Object {
      "_beagleComponent_": "custom:editableLabel",
      "_implicitContexts_": Array [
        Object {
          "id": "item",
          "value": Object {
            "color": "#ffff00",
            "id": 3,
            "name": "Supermarket",
          },
        },
        Object {
          "id": "index",
          "value": 3,
        },
      ],
      "context": Object {
        "id": "isLoading",
        "value": false,
      },
      "id": "repeater:label-list:3:0",
      "isLoading": false,
      "key": 3,
      "label": Object {
        "color": "#ffff00",
        "id": 3,
        "name": "Supermarket",
      },
      "onRemove": [Function],
      "onSave": [Function],
    },
  ],
  "dataSource": Array [
    Object {
      "color": "#00ff00",
      "id": 0,
      "name": "Personal",
    },
    Object {
      "color": "#0000ff",
      "id": 1,
      "name": "Work",
    },
    Object {
      "color": "#ff0000",
      "id": 2,
      "name": "Urgent",
    },
    Object {
      "color": "#ffff00",
      "id": 3,
      "name": "Supermarket",
    },
  ],
  "id": "label-list",
  "key": "id",
  "style": Object {
    "backgroundColor": "#333",
    "borderRadius": "20px",
    "flexDirection": "column",
    "maxWidth": "500px",
    "padding": "15px",
  },
  "template": Object {
    "_beagleComponent_": "custom:editableLabel",
    "context": Object {
      "id": "isLoading",
      "value": false,
    },
    "isLoading": "@{isLoading}",
    "label": "@{item}",
    "onRemove": Array [
      Object {
        "_beagleAction_": "beagle:condition",
        "condition": "@{isEmpty(item.id)}",
        "onFalse": Array [
          Object {
            "_beagleAction_": "beagle:setContext",
            "contextId": "isLoading",
            "value": true,
          },
          Object {
            "_beagleAction_": "beagle:sendRequest",
            "method": "DELETE",
            "onError": Array [
              Object {
                "_beagleAction_": "custom:feedback",
                "message": "Error while deleting label.",
                "type": "error",
              },
              Object {
                "_beagleAction_": "beagle:setContext",
                "contextId": "isLoading",
                "value": false,
              },
            ],
            "onSuccess": Array [
              Object {
                "_beagleAction_": "beagle:setContext",
                "contextId": "labels",
                "path": "current",
                "value": "@{remove(labels.current, item)}",
              },
            ],
            "url": "https://keep.beagle.io/label/@{item.id}",
          },
        ],
        "onTrue": Array [
          Object {
            "_beagleAction_": "beagle:setContext",
            "contextId": "labels",
            "path": "new",
          },
        ],
      },
    ],
    "onSave": Array [
      Object {
        "_beagleAction_": "beagle:setContext",
        "contextId": "isLoading",
        "value": true,
      },
      Object {
        "_beagleAction_": "beagle:sendRequest",
        "data": "@{onSave.value}",
        "method": "@{condition(isEmpty(item.id), 'POST', 'PUT')}",
        "onError": Array [
          Object {
            "_beagleAction_": "custom:feedback",
            "message": "@{error.data.message}",
            "type": "error",
          },
          Object {
            "_beagleAction_": "beagle:setContext",
            "contextId": "isLoading",
            "value": false,
          },
        ],
        "onSuccess": Array [
          Object {
            "_beagleAction_": "beagle:condition",
            "condition": "isEmpty(item.id)",
            "onFalse": Array [
              Object {
                "_beagleAction_": "beagle:setContext",
                "contextId": "labels",
                "path": "current[@{index}]",
                "value": "@{onSuccess.data}",
              },
              Object {
                "_beagleAction_": "beagle:setContext",
                "contextId": "isLoading",
                "value": false,
              },
            ],
            "onTrue": Array [
              Object {
                "_beagleAction_": "beagle:setContext",
                "contextId": "labels",
                "value": Object {
                  "current": "@{insert(labels.current, onSuccess.data)}",
                },
              },
            ],
          },
        ],
        "url": "https://keep.beagle.io/label",
      },
    ],
  },
}
`;
